src/main/java/io/aiven/kafka/connect/transforms/FilterByFieldValue.java

```diff
@@ -0,0 +1,196 @@
+            text = Optional.of(fieldValue.toString());
+        }
+        return text;
+    }
+
+    @SuppressWarnings("unchecked")
+    private R applySchemaless(final R record) {
+        if (fieldName == null || fieldName.isEmpty()) {
+            final Optional<String> value = getSchemalessFieldValue(operatingValue(record));
+            return filterCondition.test(value) ? record : null;
+        } else {
+            final Map<String, Object> map = (Map<String, Object>) operatingValue(record);
+            final Optional<String> fieldValue = getSchemalessFieldValue(map.get(fieldName));
+            return filterCondition.test(fieldValue) ? record : null;
+        }
+    }
+
+    private Optional<String> getSchemalessFieldValue(final Object fieldValue) {
+        if (fieldValue == null) {
+            return Optional.empty();
+        }
+        Optional<String> text = Optional.empty();
+        if (isSupportedType(fieldValue)) {
+            text = Optional.of(fieldValue.toString());
+        }
+        return text;
+    }
+
+    private boolean isSupportedType(final Object fieldValue) {
+        final Set<Class<?>> supportedTypes = new HashSet<>(
+            Arrays.asList(
+                String.class, Long.class, Integer.class, Short.class,
+                Double.class, Float.class, Boolean.class
+            )
+        );
+
+        return supportedTypes.contains(fieldValue.getClass());
+    }
+
+    @Override
+    public ConfigDef config() {
+        return new ConfigDef()
+            .define("field.name",
+                ConfigDef.Type.STRING,
+                null,
+                ConfigDef.Importance.HIGH,
+                "The field name to filter by."
+                    + "Schema-based records (Avro), schemaless (e.g. JSON), and raw values are supported."
+                    + "If empty, the whole key/value record will be filtered.")
+            .define("field.value",
+                ConfigDef.Type.STRING,
+                null,
+                ConfigDef.Importance.HIGH,
+                "Expected value to match. Either define this, or a regex pattern")
+            .define("field.value.pattern",
+                ConfigDef.Type.STRING,
+                null,
+                ConfigDef.Importance.HIGH,
+                "The pattern to match. Either define this, or an expected value")
+            .define("field.value.matches",
+                ConfigDef.Type.BOOLEAN,
+                true,
+                ConfigDef.Importance.MEDIUM,
+                "The filter mode, 'true' for matching or 'false' for non-matching");
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        final AbstractConfig config = new AbstractConfig(config(), configs);
+        this.fieldName = config.getString("field.name");
+        this.fieldExpectedValue = Optional.ofNullable(config.getString("field.value"));
```

Comment:
It might be useful to allow users to match more than only primitive values. We could use the [Values class](https://kafka.apache.org/34/javadoc/org/apache/kafka/connect/data/Values.html) to parse this property which would allow us to read in more complex types. In addition, this would save unnecessary calls to `toString` on field values.

This also follows the precedent set in other transforms like `InsertHeader` and how it [parses its "value.literal" property](https://github.com/apache/kafka/blob/6e8d0d9850b05fc1de0ceaf77834e68939f782c1/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertHeader.java#L75).

Reply:
updated the getStructFieldValue and getSchemalessFieldValue methods to parse more complex types using the Values class
cc @jeqo

Reply:
Thanks for the suggestion! Didn't know about Values class, looks like a great match here. 
I guess the only potential issue I foresee is when there is a mismatch between parsed schema and the desired one, e.g. if "123" is used as expected value, the inferred schema is INT8, when you may want STRING. Though, to start with this looks good to me. If an issue arise around this we could add a config for expected type.